xhistogram/core.py

```diff
@@ -332,17 +378,59 @@ def reshape_input(a):
             )
     else:
         bins = [
-            np.histogram_bin_edges(a, b, r, weights_reshaped)
-            for a, b, r in zip(all_args_reshaped, bins, range)
+            np.histogram_bin_edges(a, b, r, weights)
+            for a, b, r in zip(all_arrays, bins, range)
```

Comment:
I brought this code in when I merged with upstream/master and picked up the new bin edges logic introduced by @dougiesquire in #45. Unfortunately the tests are going to fail because of shape mismatches between arrays and weights. Due to the refactor in this PR, the high-level `histogram` function no longer has access to the reshaped / broadcasted arrays. So we will need some new solution here.

Reply:
Maybe it's cleanest/easiest then to `np.broadcast_arrays(*all_arrays)` early on as @gjoseph92 suggests in https://github.com/xgcm/xhistogram/pull/49#discussion_r640872588? Then we also don't need to worry about each array potentially having a different `input_index`.

Another option I guess would be to do the broadcasting in xarray.histogram and have core.histogram expect that all `args` and `weights` are the same shape (somewhat replicating what numpy.histogram expects).

Reply:
@rabernat it - the docstring of `np.histogram_bin_edges` says that the weights argument "is currently not used by any of the bin estimators, but may be in the future." So it currently doesn't do anything: you can make all the tests pass again simply by deleting the weights argument from the call to `np.histogram_bin_edges`.

Separately I do think that we should consider broadcasting early on to make the indexing easier to follow, but it would be nice to get this PR merged and discuss that refactoring elsewhere.